# SPAD
This set of codes is used to call Speckle Associated Domains (SPADs), compare SPADs from different cell lines, and correlate with gene expression.

## Percentile normalization
SPADs are defined as genomic regions with top 5 percentile SON TSA-Seq scores. So we first convert TSA-Seq scores into percentiles.

```shell
python TSA_percentile_norm_TSA2.0.py -w TSA-Seq_hanning_20kbx21.wig -q 100 -g utilities/hg38_Gap.bed -o TSA-Seq_hanning_20kbx21_percentile -gg utilities/hg38F.genome
#Genome size file hg38F.genome was for female cell line (K562), hg38M.genome was for male cell lines (H1, HCT116, HFFc6).
```

Figure 2a (middle) was generated from the bigwig file (TSA-Seq_hanning_20kbx21_percentile.bw) by this code.

## SPADs calling
Use the percentile wig files from last step, identify bins above 95 percentile and merge adjacent bins to call SPADs

```shell
python BigPercentiles_TSA2.0.py -w TSA-Seq_hanning_20kbx21_percentile.wig -o TSA-Seq_hanning_20kbx21_percentile -p 95 -g utilities/hg38F.genome -win 20000
#Genome size file hg38F.genome was for female cell line (K562), hg38M.genome was for male cell lines (H1, HCT116, HFFc6).
```
This code will generate a TSA-Seq_hanning_20kbx21_percentile_above_95.0.bed file for all 20-kb bins above 95 percentiles and a TSA-Seq_hanning_20kbx21_percentile_above_95.0_mergeAdjacent.bed (and corresponing bigbed) file for SPADs by merging adjacent 20-kb bins.

This code will also generate simple statistics of region size and number with histograms.

Figure 2a (bottom) was generated from bigbed file (TSA-Seq_hanning_20kbx21_percentile_above_95.0_mergeAdjacent.bb) by this code.

## Compare SPADs from different cell lines
### Take SPADs from one cell line, check region percentiles in other cell lines
Use the mergeAdjacent.bed file generated from last step, collect all regions (SPADs) and calculate region percentiles in other cell lines (input percentile wig file), and return a box plot

```shell
python checkSpadsInOtherCellLine_v2_TSA2.0.py -b cell0_TSA-Seq_hanning_20kbx21_percentile_above_95.0_mergeAdjacent.bed -w1 cell1_TSA-Seq_hanning_20kbx21_percentile.wig -w2 _cell2_TSA-Seq_hanning_20kbx21_percentile.wig -w3 cell3_TSA-Seq_hanning_20kbx21_percentile.wig -o cell0withOthers.eps
```

Figure 2b, Supplementary Figure 7a-c were generated by this code.

### 4-way Venn diagram
Use the four TSA-Seq_hanning_20kbx21_percentile_above_95.0.bed files from the four cell lines generated in "SPADs calling" step to generate a 4-way Venn diagram by a tool called Intervene (version 0.6.4, https://intervene.readthedocs.io/en/latest/).

```shell
intervene venn -i cell1_TSA-Seq_hanning_20kbx21_percentile_above_95.0.bed cell2_TSA-Seq_hanning_20kbx21_percentile_above_95.0.bed cell3_TSA-Seq_hanning_20kbx21_percentile_above_95.0.bed cell4_TSA-Seq_hanning_20kbx21_percentile_above_95.0.bed --filenames
```

Supplementary Figure 7d was generated by this code. The resulting intervened numbers are numbers for 20kb bins. Genomic size were calculated by multiply the reported number by 20kb.
